debezium-embedded/src/main/java/io/debezium/embedded/AsyncEmbeddedEngine.java

```diff
@@ -0,0 +1,1266 @@
+
+        TransformAndConsumeRecord(final SourceRecord record, final Transformations transformations, final Consumer<SourceRecord> consumer) {
+            this.record = record;
+            this.transformations = transformations;
+            this.consumer = consumer;
+        }
+
+        @Override
+        public Void call() {
+            final SourceRecord transformedRecord = transformations.transform(record);
+            if (transformedRecord != null) {
+                consumer.accept(transformedRecord);
+            }
+            return null;
+        }
+    }
+
+    /**
+     * {@link Callable} which applies transformations to a single record, transformed it into desired format and applies provided use consumer on this record.
+     */
+    private static class TransformConvertConsumeRecord<R> implements Callable<Void> {
+        private final SourceRecord record;
+        private final Transformations transformations;
+        private final Function<SourceRecord, R> serializer;
+        private final Consumer<R> consumer;
+
+        TransformConvertConsumeRecord(final SourceRecord record, final Transformations transformations, final Function<SourceRecord, R> serializer,
+                                      final Consumer<R> consumer) {
+            this.record = record;
+            this.transformations = transformations;
+            this.serializer = serializer;
+            this.consumer = consumer;
+        }
+
+        @Override
+        public Void call() {
+            final SourceRecord transformedRecord = transformations.transform(record);
+            if (transformedRecord != null) {
+                consumer.accept(serializer.apply(transformedRecord));
+            }
+            return null;
+        }
+    }
+
+    /**
+     * {@link RecordProcessor} which runs transformations of the records in parallel and then pass the whole batch to the user-provided handler.
+     * This processor should be used when user provides its own {@link DebeziumEngine.ChangeConsumer} and records shouldn't be converted to different format.
+     */
+    private class ParallelSmtBatchProcessor extends AbstractRecordProcessor {
+        final DebeziumEngine.ChangeConsumer<SourceRecord> userHandler;
+
+        ParallelSmtBatchProcessor(final DebeziumEngine.ChangeConsumer<SourceRecord> userHandler) {
+            this.userHandler = userHandler;
+        }
+
+        @Override
+        public void processRecords(final List<SourceRecord> records) throws Exception {
+            LOGGER.debug("Thread {} is submitting {} records for processing.", Thread.currentThread().getName(), records.size());
+            final List<Future<SourceRecord>> recordFutures = new ArrayList<>(records.size());
+            records.stream().forEachOrdered(r -> recordFutures.add(recordService.submit(new TransformRecord(r, transformations))));
+
+            LOGGER.debug("Thread {} is getting source records.", Thread.currentThread().getName());
```

Comment:
Here and other processors - again some of the messages might be better to change to `TRACE`